builtin/logical/ssh/path_roles.go

```diff
@@ -0,0 +1,245 @@
+	}
+	keyEntry, err := req.Storage.Get(fmt.Sprintf("keys/%s", keyName))
+	if err != nil || keyEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("Invalid 'key': '%s'", keyName)), nil
+	}
+
+	adminUser := d.Get("admin_user").(string)
+	if adminUser == "" {
+		return logical.ErrorResponse("Missing admin username"), nil
+	}
+
+	cidr := d.Get("cidr").(string)
+	if cidr == "" {
+		return logical.ErrorResponse("Missing cidr blocks"), nil
+	}
+	for _, item := range strings.Split(cidr, ",") {
+		_, _, err := net.ParseCIDR(item)
+		if err != nil {
+			return logical.ErrorResponse(fmt.Sprintf("Invalid cidr entry '%s'", item)), nil
+		}
+	}
+
+	defaultUser := d.Get("default_user").(string)
+	if defaultUser == "" {
+		defaultUser = adminUser
+	}
+
+	port := d.Get("port").(string)
+	if port == "" {
+		port = "22"
+	}
+
+	entry, err := logical.StorageEntryJSON(fmt.Sprintf("policy/%s", roleName), sshRole{
+		KeyName:     keyName,
+		AdminUser:   adminUser,
+		DefaultUser: defaultUser,
+		CIDR:        cidr,
+		Port:        port,
+		KeyType:     KeyTypeDynamic,
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := req.Storage.Put(entry); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+func (b *backend) pathRoleWrite(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	keyType := d.Get("key_type").(string)
+	if keyType == "" {
+		return logical.ErrorResponse("Missing key type"), nil
+	}
+	keyType = strings.ToLower(keyType)
+
+	if keyType == KeyTypeOTP {
+		return createOTPRole(req, d)
+	} else if keyType == KeyTypeDynamic {
+		return createDynamicKeyRole(req, d)
+	} else {
+		return logical.ErrorResponse("Invalid key type"), nil
+	}
+}
+
+func (b *backend) pathRoleRead(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("name").(string)
+	roleEntry, err := req.Storage.Get(fmt.Sprintf("policy/%s", roleName))
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, nil
+	}
+
+	var role sshRole
+	if err := roleEntry.DecodeJSON(&role); err != nil {
+		return nil, err
+	}
+
+	return &logical.Response{
+		Data: map[string]interface{}{
+			"key":          role.KeyName,
+			"admin_user":   role.AdminUser,
+			"default_user": role.DefaultUser,
+			"cidr":         role.CIDR,
+			"port":         role.Port,
+			"key_type":     role.KeyType,
+		},
+	}, nil
+}
+
+func (b *backend) pathRoleDelete(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	roleName := d.Get("name").(string)
+	err := req.Storage.Delete(fmt.Sprintf("policy/%s", roleName))
+	if err != nil {
+		return nil, err
+	}
+	return nil, nil
+}
+
+type sshRole struct {
+	KeyType     string `json:"key_type"`
+	KeyName     string `json:"key"`
+	AdminUser   string `json:"admin_user"`
+	DefaultUser string `json:"default_user"`
+	CIDR        string `json:"cidr"`
+	Port        string `json:"port"`
+}
+
+const KeyTypeOTP = "otp"
```

Comment:
Can you move the type definitions and const to the top of the file? Leave the help at the bottom